### PR TITLE
Extends true_divide to be a method

### PR DESCRIPTION
**Pull request by mruberry**
Per title. See related https://github.com/pytorch/pytorch/pull/34570.

In PyTorch 1.7 the plan is for torch.div and Python's division operator to perform "true" division, like Python 3, JAX, and NumPy. To facilitate this change, this PR expands true_divide to be a method so it can cover all of torch.div's use cases. 

New true_divide tests are added to test_torch.py, test_type_promotion.py, and test_sparse.py.

### CODE REVIEW

File: test/test_torch.py [hunk 1]
```diff
@@ -6002,6 +6002,59 @@ def test_conv_transposed_backward_agnostic_to_memory_format(self, device):
...
+        self.assertTrue(torch.div is torch.true_divide)
+        self.assertTrue(torch.Tensor.div is torch.Tensor.true_divide)
+        self.assertTrue(torch.Tensor.div_ is torch.Tensor.true_divide_)
+        self.assertTrue(torch.Tensor.__truediv__ is torch.Tensor.true_divide)
+
+        # Tests dense tensor behavior
+        t = torch.tensor((2,), device=device, dtype=torch.int)
+        d = torch.tensor((1,), device=device, dtype=torch.int)
+        o = torch.empty((2,), device=device, dtype=torch.int)
+        self.assertTrue((t / d).dtype, torch.get_default_dtype())
```
Comment (gchanan) [inline test/test_torch.py]: I think you need to make this work with python 2.

File: torch/future_div.py [hunk 2]
```diff
@@ -0,0 +1,6 @@
+import torch
+
+torch.div = torch.true_divide
```
Comment (gchanan) [inline torch/future_div.py]: there's some existing thing like torch.future or torch.__future__ already, let's put it there.

File: test/test_torch.py (hunk 1 above)
Comment (mruberry) [reply test/test_torch.py]: This test is gone so it's moot.

File: torch/future_div.py (hunk 2 above)
Comment (mruberry) [reply torch/future_div.py]: This is moot. We removed the future div stuff.

File: test/test_type_promotion.py [hunk 3]
```diff
@@ -442,25 +442,42 @@ def test_transpose(self, device):
     @dtypes(torch.bool, torch.short, torch.uint8, torch.int, torch.long)
     @float_double_default_dtype
     def test_true_divide(self, device, dtype):
-        dividend = torch.randn(5, device=device).to(dtype)
+        dividend = (torch.randn(5, device=device) * 100).to(dtype)
```
[the remark this answers is not in the document]
Comment (mruberry) [reply test/test_type_promotion.py]: To make the division more interesting. Example dividends:

```
(torch.randn(100)).long()
: tensor([ 0,  0, -1, -2,  0,  0,  1,  0,  0,  0,  0,  0,  0,  0,  0,  0, -1,  0,
         1, -1,  0, -1,  0,  0,  0,  0,  1,  0,  0,  0, -1,  1,  0,  0,  1,  0,
         0,  1,  0,  0,  0,  0, -1, -1,  0, -1,  0,  0,  0, -1,  0,  0, -2,  0,
         0,  0,  1, -1,  0,  0, -1,  0,  0,  0,  0, -1,  1,  0,  0,  0, -1, -1,
         0,  0, -1, -1,  0,  0,  0,  2,  0,  1,  0,  0,  0,  0,  0,  1,  0,  0,
         0,  0,  0,  1, -1,  0,  0,  1,  0,  1])

(torch.randn(100) * 100).long()
: tensor([  54,  113,    7,   69,    4,   98,  -32,  223,  -36,  285, -111,   -6,
         -36,  -43,  122,  -39, -164,  101, -100, -122,   -2,  -70,   40,  121,
          16,   21,    6,   21, -110,  -35, -186,  193,   15,  -21,   75,   61,
         -62,  -27, -116,  -32,  -17, -159, -111,   56,  -28,   59, -117,   63,
         -57,  -49,  210,   93,   11, -107,  -66,   32,   -8, -150,  -20,   46,
           6,   98,  -28,   87,    6, -124,  -81,   42,  -68,  -14,  -39, -149,
         117,   48,  -81, -120, -138,  183,  146,  -43,  -72,  -53,  -93,   67,
         -73, -161,  152,   57, -229,  -37,   59,  -85,   47,  178, -127, -110,
          34,  175,  -11,  -39])
```

File: test/test_type_promotion.py [hunk 4]
```diff
@@ -442,25 +442,42 @@ def test_transpose(self, device):
...
         divisor = torch.arange(1, 6, device=device).to(dtype)
         casting_result = dividend.to(torch.get_default_dtype()) / divisor.to(torch.get_default_dtype())
         self.assertEqual(casting_result, torch.true_divide(dividend, divisor))
 
-    @dtypes(torch.bool, torch.short, torch.uint8, torch.int, torch.long)
+    @dtypes(torch.float, torch.double, torch.bool, torch.short, torch.uint8, torch.int, torch.long)
     def test_true_divide_out(self, device, dtype):
-        dividend = torch.randn(5, device=device).to(dtype)
+        dividend = (torch.randn(5, device=device) * 100).to(dtype)
```
[the remark this answers is not in the document]
Comment (mruberry) [reply test/test_type_promotion.py]: See note above.

File: test/test_torch.py [hunk 5]
```diff
@@ -6002,6 +6002,59 @@ def test_conv_transposed_backward_agnostic_to_memory_format(self, device):
...
+        self.assertTrue((s / 2).dtype, torch.get_default_dtype())
+        self.assertTrue(torch.div(s, 2).dtype, torch.get_default_dtype())
+        self.assertTrue(s.div(2).dtype, torch.get_default_dtype())
+        with self.assertRaises(RuntimeError):
+            s.div_(2)
+        with self.assertRaises(RuntimeError):
+            torch.div(s, 2, out=o)
+
+        # Undoes torch.future_div import
+        torch.div = prev_div
```
[the remark this answers is not in the document]
Comment (mruberry) [reply test/test_torch.py]: This test doesn't exist anymore.

File: test/test_torch.py [hunk 6]
```diff
@@ -6002,6 +6002,59 @@ def test_conv_transposed_backward_agnostic_to_memory_format(self, device):
...
+        import importlib
+        importlib.reload(torch.future_div)
+
+        self.assertTrue(torch.div is torch.true_divide)
+        self.assertTrue(torch.Tensor.div is torch.Tensor.true_divide)
+        self.assertTrue(torch.Tensor.div_ is torch.Tensor.true_divide_)
+        self.assertTrue(torch.Tensor.__truediv__ is torch.Tensor.true_divide)
+
+        # Tests dense tensor behavior
+        t = torch.tensor((2,), device=device, dtype=torch.int)
```
[the remark this answers is not in the document]
Comment (mruberry) [reply test/test_torch.py]: Moot.

File: test/test_torch.py [hunk 7]
```diff
@@ -6002,6 +6002,59 @@ def test_conv_transposed_backward_agnostic_to_memory_format(self, device):
...
+        d = torch.tensor((1,), device=device, dtype=torch.int)
+        o = torch.empty((2,), device=device, dtype=torch.int)
+        self.assertTrue((t / d).dtype, torch.get_default_dtype())
+        self.assertTrue((t / 2).dtype, torch.get_default_dtype())
+        self.assertTrue(torch.div(t, d).dtype, torch.get_default_dtype())
```
[the remark this answers is not in the document]
Comment (mruberry) [reply test/test_torch.py]: This test is dead, though.

File: test/test_type_promotion.py [hunk 8]
```diff
@@ -442,26 +442,57 @@ def test_transpose(self, device):
...
         divisor = torch.arange(1, 6, device=device).to(dtype)
 
         # Tests that requests for an integer quotient fail
-        integral_quotient = torch.empty(5, device=device, dtype=dtype)
-        with self.assertRaises(RuntimeError):
-            torch.true_divide(dividend, divisor, out=integral_quotient)
+        if dtype in (torch.bool, torch.short, torch.uint8, torch.int, torch.long):
```
Comment (gchanan) [inline test/test_type_promotion.py]: super nit: it's slightly nicer if you do all bit-length variants, easier to read (since you have to for uint8).

**Review by gchanan — APPROVED** [review]

File: test/test_type_promotion.py (hunk 8 above)
Comment (mruberry) [reply test/test_type_promotion.py]: Super nits deserve love, too.

Comment (mruberry) [reply test/test_type_promotion.py]: Fixed.

**Pull request closed without merging**